src/string/CreditCard.ts

```diff
@@ -0,0 +1,168 @@
+import { flow, pipe } from 'fp-ts/function'
+import * as fc from 'fast-check'
+import * as RA from 'fp-ts/ReadonlyArray'
+
+import * as Arb from '../internal/ArbitraryBase'
+import { MonoidSum } from 'fp-ts/lib/number'
+import { digits } from '../internal/util'
+
+/**
+ * @since 0.0.3
+ * @category Internal
+ */
+interface CreditCardBrand {
+  readonly CreditCard: unique symbol
+}
+
+/**
+ * TODO: Add module comment
+ *
+ * @since 0.0.3
+ * @category Model
+ */
+export type CreditCard = string & CreditCardBrand
+
+/**
+ * @since 0.0.3
+ * @category Model
+ */
+export type SchemableParams<S> = HKT2<S, string, CreditCard>
+
+/**
+ * @since 0.0.3
+ * @category Model
+ */
+export type SchemableParams1<S extends URIS> = Kind<S, CreditCard>
+
+/**
+ * @since 0.0.3
+ * @category Model
+ */
+export type SchemableParams2<S extends URIS2> = Kind2<S, string, CreditCard>
+
+/**
+ * @since 0.0.3
+ * @category Model
+ */
+export type SchemableParams2C<S extends URIS2> = Kind2<S, unknown, CreditCard>
+
+const cc =
+  /(^4[0-9]{12}(?:[0-9]{3})?$)|(^(?:5[1-5][0-9]{2}|222[1-9]|22[3-9][0-9]|2[3-6][0-9]{2}|27[01][0-9]|2720)[0-9]{12}$)|(3[47][0-9]{13})|(^3(?:0[0-5]|[68][0-9])[0-9]{11}$)|(^6(?:011|5[0-9]{2})[0-9]{12}$)|(^(?:2131|1800|35\d{3})\d{11}$)/
+
+/** This calculates the Luhn checksum for a given string of digits. */
+const luhn: (cc: string) => number = flow(
+  Str.split(''),
+  RA.reverse,
+  RA.map(d => parseInt(d)),
+  RA.foldMapWithIndex(MonoidSum)((i, d) =>
+    i % 2 === 1
+      ? // numbers of odd rank stay the same
+        d
+      : // numbers of even rank get doubled, then their digits are summed
+      d * 2 >= 10
+      ? 1 + (d * 2 - 10)
+      : d * 2
+  ),
+  sum => (sum % 10) % 10,
+  checksum => (checksum === 0 ? 0 : 10 - checksum)
+)
+
+/**
+ * @since 0.0.3
+ * @category Refinements
+ */
+export const isCreditCard = (s: string): s is CreditCard => {
+  const clean = s.replace(/[^0-9]/g, '')
+  return cc.test(clean) && luhn(clean.slice(0, -1)) === parseInt(clean.slice(-1))
+}
+
+/**
+ * @since 0.0.3
+ * @category Instances
+ */
+export const Decoder: SchemableParams2C<D.URI> = pipe(
+  D.string,
+  D.refine(isCreditCard, 'CreditCard')
+)
+
+/**
+ * @since 0.0.3
+ * @category Instances
+ */
+export const Encoder: SchemableParams2<Enc.URI> = Enc.id()
+
+/**
+ * @since 0.0.3
+ * @category Instances
+ */
+export const Eq: SchemableParams1<Eq_.URI> = Str.Eq
+
+/**
+ * @since 0.0.3
+ * @category Instances
+ */
+export const Guard: SchemableParams1<G.URI> = pipe(G.string, G.refine(isCreditCard))
+
+/**
+ * @since 0.0.3
+ * @category Instances
+ */
+export const TaskDecoder: SchemableParams2C<TD.URI> = pipe(
+  TD.string,
+  TD.refine(isCreditCard, 'CreditCard')
+)
+
+/**
+ * @since 0.0.3
+ * @category Instances
+ */
+export const Type: SchemableParams1<t.URI> = pipe(
+  t.string,
+  t.refine(isCreditCard, 'CreditCard')
+)
+
+/**
+ * @since 0.0.3
+ * @category Instances
+ */
+export const Arbitrary: SchemableParams1<Arb.URI> = fc
+  .oneof(
```

Comment:
😅 I worry a little bit about `fast-check` blowing up the size of the library. It totally seems worth the cost since these Arbitraries are unreasonably useful.. but we should probably also consider impact on bundle size. The only alternative I can think of right now is splitting the Arbitraries out into their own package.. and I don't think it should be a concern right now, I'm just publicly airing my neuroses.

Reply:
I thought about that. Tree shaking should take care of it though, right? This is a library after all 🙂